src/telemetry/ActivityMeasurementService.ts

```diff
@@ -0,0 +1,119 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.md in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import { ext } from '../extensionVariables';
+import { AsyncLazy } from '../utils/lazy';
+
+const defaultMeasurement: ActivityMeasurement = {
+    lastSession: undefined,
+    monthlySessions: 0,
+    totalSessions: 0,
+};
+
+export type ActivityType = 'overall' | 'overallnoedit';
+
+export interface IActivityMeasurementService {
+    recordActivity(type: ActivityType): Promise<void>;
+    getActivity(type: ActivityType): ActivityMeasurement;
+}
+
+export interface ActivityMeasurement {
+    lastSession: number | undefined;
+    monthlySessions: number;
+    totalSessions: number;
+}
+
+export class ActivityMeasurementService implements IActivityMeasurementService {
+    private readonly lazySetterMap: Map<ActivityType, AsyncLazy<void>> = new Map<ActivityType, AsyncLazy<void>>();
+    private readonly values: Map<ActivityType, ActivityMeasurement> = new Map<ActivityType, ActivityMeasurement>();
+
+    public constructor(private readonly memento: vscode.Memento) {
+    }
+
+    /**
+     * Records activity. Once per session (max 1 per day), it will increment the monthly and total session counts and set the date of the last session to now.
+     * Calling with any type will also result in 'overall' activity being incremented, and any except edit will increment 'overallnoedit'.
+     * @param type The activity type to record measurements for
+     */
+    public async recordActivity(type: ActivityType): Promise<void> {
+        if (!ext.telemetryOptIn) {
+            return;
+        }
+
+        try {
+            if (!this.lazySetterMap.has(type)) {
+                this.lazySetterMap[type] = new AsyncLazy(async () => {
+                    const currentValue = this.getActivity(type);
+                    const now = Date.now();
+
+                    // No need to increment if it's been done already today
+                    if (sameDate(currentValue.lastSession, now, 'day')) {
+                        return;
+                    }
+
+                    const newValue: ActivityMeasurement = {
+                        lastSession: now,
+                        monthlySessions: currentValue.monthlySessions + 1,

```

Comment:
The name "monthly" implied to me that we'll record 1 instance of ActivityMeasurement per month, which isn't really the case. So I suggest renaming it to something like "currentMonthSessions" or something similar to avoid confusion.